flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+
+    private final AsyncExceptionHandler asyncExceptionHandler;
+
+    private final int allowedNumberOfFailures;
+
+    /** Materialization failure retries. */
+    private final AtomicInteger retries;
+
+    /** Making sure only one materialization on going at a time. */
+    private final AtomicBoolean materializationOnGoing;
+
+    private long materializedId;
+
+    private MaterializedState materializedState;
+
+    public PeriodicMaterializer(
+            MailboxExecutor mailboxExecutor,
+            ExecutorService asyncOperationsThreadPool,
+            Snapshotable keyedStateBackend,
+            CheckpointStorageWorkerView checkpointStorageWorkerView,
+            StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter,
+            AsyncExceptionHandler asyncExceptionHandler,
+            MaterializedState materializedState,
+            long periodicMaterializeInitDelay,
+            long periodicMaterializeInterval,
+            int allowedNumberOfFailures,
+            boolean materializationEnabled) {
+        this.mailboxExecutor = mailboxExecutor;
+        this.asyncOperationsThreadPool = asyncOperationsThreadPool;
+        this.keyedStateBackend = keyedStateBackend;
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.asyncExceptionHandler = asyncExceptionHandler;
+        this.periodicExecutor =
+                Executors.newSingleThreadScheduledExecutor(
+                        new ExecutorThreadFactory("periodic-materialization"));
+        this.streamFactory = shared -> checkpointStorageWorkerView.createTaskOwnedStateStream();
+        this.allowedNumberOfFailures = allowedNumberOfFailures;
+        this.materializationOnGoing = new AtomicBoolean(false);
+        this.retries = new AtomicInteger(allowedNumberOfFailures);
+
+        this.materializedId = 0;
+        this.materializedState = materializedState;
+
+        if (materializationEnabled) {
+            this.periodicExecutor.scheduleAtFixedRate(
+                    this::triggerMaterialization,
+                    periodicMaterializeInitDelay,
+                    periodicMaterializeInterval,
+                    TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // Only one materialization ongoing at a time
+                    if (!materializationOnGoing.compareAndSet(false, true)) {
+                        return;
+                    }
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+                    if (upTo.equals(materializedState.lastMaterializedTo())) {
+                        return;
+                    }
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
```

Comment:
Should we handle any possible exceptions here to honour retry limit?

Reply:
Do you mean the sync phase? In the sync phase, if it fails, the task should fail?

Or do we want to allow sync phase to fail as well? Can you give me an example that the sync phase fails but we should tolerate that?

Reply:
Yes, I mean the sync phase. I think recoverable I/O failures are possible here, but I'm not sure if they should be tolerated TBH.

For example: timeout writing to SAN/NAS, no disk space, file handle limit exceeded, bad sectors not handled by the FS.

If it's too difficult to honour the limit here then failover is also an option.

Reply:
The problem is that most of the sync failure is non-tolerable/recoverable (or do not know whether we can tolerate or recovered). It is difficult to distinguish all these cases.